### ISSUE TITLE
ObjC: Error when using a certificate chain that's not the default one

### ISSUE THREAD

**Issue by jclardy** [grpc/grpc#3363]
I am trying to connect to our secure domain using the Objective-C library but I can't find any way to specify my own client TLS certificates. Is this supported already? I can't find anything for it.

I see that roots.pem is loaded from the grpc bundle and I tried just appending my certs to that file but it didn't seem to work, and either way I would have to redo this on every machine the code is built on. Is there something I'm missing in the podspec file?

**Comment by jcanizales** [comment]
Hi,

if I'm understanding your use case correctly, this isn't currently supported in the ObjC library. But let's double check: you're trying to make a gRPC connection from a Mac (*not* iOS) to another server, right?

**Comment by jclardy** [comment]
Sorry I wasn't clear, I am building an iOS app and am trying to specify the TLS certificate inside that app.

**Comment by jcanizales** [comment]
No worries! And the certificate you're trying to use identifies the server (your secure domain), right? *Not* the client app, correct?

**Comment by jclardy** [comment]
Yes that is correct

**Comment by jcanizales** [comment]
Ok then, we're on the same page now :)

> I tried just appending my certs to that file but it didn't seem to work

This should have worked, even if it's not an ideal solution... Is your cert on PEM format? What did the library output in the logs console when trying to connect? What's the result if you replace the whole contents of the file with your cert?

The ideal solution would be to use [this method](https://github.com/grpc/grpc/blob/master/src/objective-c/tests/InteropTestsLocalSSL.m#L57), which was created for calling test servers. But if adding your cert to the default file doesn't work, this shouldn't either!

**Comment by jclardy** [comment]
I'm using the library via cocoapods, it seems that there are updates in the repo that aren't in the cocoapods version (Also I can't access the private headers from my app.)

But when I replaced the roots.pem with my cert I got this error:
```
E0918 11:35:40.167109000 2953924608 security_connector.c:447] Missing selected ALPN property.
E0918 11:35:40.167695000 2953924608 secure_transport_setup.c:118] Peer check failed.
E0918 11:35:40.167829000 2953924608 secure_channel_create.c:82] Secure transport setup failed with error 2.
```

Is there some server configuration that is needed to make this work? I also noticed differences in the files on the repo from the 0.6 version on cocoapods, in things like GRPCSecureChannel. Should I try using the source straight from the repo?

**Comment by jcanizales** [comment]
Oh, I see. Yes, I'm close to sending the next release to Cocoapods (the last one with breaking changes :) ), so please clone (or submodule) this repo and tell Cocoapods to use it instead of v0.6 (and to use Protobuf version "3.0.0-alpha-4.1"). To do that, add the following lines to your Podfile:

```ruby
pod 'Protobuf', "3.0.0-alpha-4.1"
pod 'gRPC', :path => "path/to/the/root/of/this/repo"
```

This way we won't be debugging old code.

**Comment by jclardy** [comment]
I actually am working on doing that now, but now I am getting a crash on a mutex lock. I built and installed the grpc-objc plugin from source as the one I installed via homebrew seemed to compile the old version.

**Comment by jcanizales** [comment]
Ah, yes! Sorry, that's Issue https://github.com/grpc/grpc/issues/3364. If you checkout the repo before commit https://github.com/grpc/grpc/commit/131f6edafa1a281604fffc24019dcb99134eec9f it doesn't happen.

**Comment by jclardy** [comment]
```E0922 15:48:17.435238000 2953924608 security_connector.c:451] Missing selected ALPN property.
E0922 15:48:17.435632000 2953924608 secure_transport_setup.c:123] Peer check failed.
E0922 15:48:17.435642000 2953924608 secure_channel_create.c:85] Secure transport setup failed with error 2.```

Ok so I am now on version 0.7.0 (The commit before the one you posted), but still getting this error. We have the x509 cert from the server when we are in ssl_handshaker_extract_peer but the alpn and npn methods return null. 

The server is using grpc-go I believe, and we have a go based client that is able to connect to it. This is the code they are using:

```
	creds, err := credentials.NewClientTLSFromFile("***certname***.cert", "***domain_name***")
	if err != nil {
		log.Fatalf("> failed to create transport authenticator: %s", err)
	}

	conn, err := grpc.Dial(routerEndpoint, grpc.WithTransportCredentials(creds))
```
They basically just have to set the client TLS certificate and domain, then connect with Dial.

Basically using our old stack the iOS app could connect and work, but the new stack requires TLS and I can no longer connect.

**Comment by jcanizales** [comment]
It _might_ be that this `.cert` file can't replace the default `certs.pem` file...

Hey @jboeuf, do you know if a `.cert` file with an x509 cert can work as a substitute of the `certs.pem` file we use by default? Or are they completely different things?

**Comment by jclardy** [comment]
@jcanizales Do you have an estimate on when the cocoapods version will be updated (Just a general guess is fine, next few weeks or longer?) And will that be with version 0.11? Or an earlier one?

**Comment by jboeuf** [comment]
Sorry I dropped the ball on this one...

@jcanizales Indeed, an X509 PEM (which is text/base64 as opposed to DER which is binary) encoded cert can be used as a substitute for our roots.pem. roots.pem is just a list of PEM encoded certs.

@jclardy It looks like (90% sure) your issue does not come from the cert but from your server TLS implementation that does not support ALPN or NPN (needed for an HTTP2 stack like gRPC):
```
E0918 11:35:40.167109000 2953924608 security_connector.c:447] Missing selected ALPN property.
E0918 11:35:40.167695000 2953924608 secure_transport_setup.c:118] Peer check failed.
E0918 11:35:40.167829000 2953924608 secure_channel_create.c:82] Secure transport setup failed with error 2.
```
Here, the TLS handshake actually worked OK (the server cert was validated with your supplied root) but your server did not send its ALPN or NPN version. Which server are you using? A server from the gRPC package?

**Comment by jcanizales** [comment]
@jclardy I'm actually submitting v. 0.11 to the Cocoapods main repo either today or early next week :)

**Comment by jcanizales** [comment]
@jboeuf They have a go client that's connecting successfully, so the fingers point at something wrong in the client.

**Comment by jboeuf** [comment]
Maybe the go client does not enforce ALPN...

    Julien.

**Comment by jcanizales** [comment]
But then this would show up on our interop tests, and e.g. the C client and Ruby client (which wraps C, like the Objective-C one) are currently communicating successfully with a Go server.

**Comment by jboeuf** [comment]
Is the server our Go server?

**Comment by jcanizales** [comment]
@jclardy can you find out how's your server being configured exactly?

**Comment by jclardy** [comment]
@jcanizales @jboeuf it is the Go one, do the interop tests include tls connections? We had a working unsecured connection before moving to TLS.

**Comment by jboeuf** [comment]
Interesting. I'm pretty sure they do. @yang-g will know for sure: 
Yang, do we have end-to-end test with TLS from a C-based client to a Go server?

**Comment by yang-g** [comment]
We have interop tests with tls between all c-based clients to go server.

**Comment by jcanizales** [comment]
> all c-based clients to go server

Except the Objective-C one :) because the infrastructure is completely different. But it still releases the Go server from suspicion.

@jclardy At this point I can just think of further tests to make, to investigate further:
* What's the error in the logs if you *don't* replace the default roots.pem with your certificate?
* And if you *append* your certificate to the end of it?
* And if you do the above but connecting to `grpc-test.sandbox.google.com` instead of your server?

**Comment by jboeuf** [comment]
@jclardy Also, with the current failure, can you put a breakpoint in ssl_check_peer (in security_connector.c) and give me the values in the peer?

**Comment by jboeuf** [comment]
@jcanizales When these logs are written, the certificate chaining check should have already passed (unless if there's a bug in our code).

**Comment by jcanizales** [comment]
Yeah, if he can't connect to `grpc-test.sandbox.google.com` with the default certs, then we have proof the problem is unrelated to his using a different cert, and we can try and debug there.

**Comment by jcanizales** [comment]
Any luck with these tests, @jclardy?

**Comment by pquerna** [comment]
@jclardy @jcanizales see #5592 for setting client certificates (tls mutual authentication)

**Comment by jcanizales** [comment]
Hi @jclardy, were you ever able to solve the problem?

**Comment by hsaliak** [comment]
Removing GA label on this, but keeping the bug open.

**Comment by makdharma** [comment]
Closing for lack of communication.
